web/src/main/java/org/springframework/security/web/header/writers/HpkpHeaderWriter.java

```diff
@@ -0,0 +1,268 @@
+/*
+ * Copyright 2002-2016 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.web.header.writers;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.springframework.security.web.header.HeaderWriter;
+import org.springframework.security.web.util.matcher.RequestMatcher;
+import org.springframework.util.Assert;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.net.URI;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+/**
+ * Provides support for <a href="http://tools.ietf.org/html/rfc7469">HTTP Public Key Pinning (HPKP)</a>.
+ *
+ * <p>
+ * Since <a href="http://tools.ietf.org/html/rfc7469#section-4.1">Section 4.1</a> states
+ * that a value on the order of 60 days (5,184,000 seconds) may be considered a good balance,
+ * we use this value as the default. This can be customized using {@link #setMaxAgeInSeconds(long)}.
+ * </p>
+ *
+ * <p>
+ * Because <a href="http://tools.ietf.org/html/rfc7469#appendix-B">Appendix B</a> recommends
+ * that operators should first deploy public key pinning by using the report-only mode,
+ * we opted to use this mode as default. This can be customized using {@link #setReportOnly(boolean)}.
+ * </p>
+ *
+ * <p>
+ * Since we need to validate a certificate chain, the "Public-Key-Pins" or "Public-Key-Pins-Report-Only" header
+ * will only be added when {@link HttpServletRequest#isSecure()} returns {@code true}.
+ * </p>
+ *
+ * @author Tim Ysewyn
+ * @since 4.1
+ */
+public final class HpkpHeaderWriter implements HeaderWriter {
```

Comment:
@TYsewyn Since only sha256 pin is currently supported perhaps we should have a method that allows adding a (or a set of) sha256 pin(s)?

We might also provide details on how to calculate the hash (or a link to something rather permanent that describes how).